File: Miniproject-main/Miniproject-main/backend/data/generate_expanded_data.js
const fs = require('fs');
const path = require('path');

function generateDataset() {
    const filename = path.join(__dirname, 'comprehensive_transactions_100.csv');
    const headers = [
        "TransactionID", "AccountID", "AccountHolder", "TransactionType",
        "TransactionDate", "Amount", "AccountBalance", "MerchantID",
        "Location", "SenderName", "SenderID", "SenderAccountNo"
    ];

    const rows = [headers.join(',')];
    let start_date = new Date(2023, 10, 1, 9, 0, 0); // Nov 1st

    const users = ["Alice", "Bob", "Charlie", "David", "Eve", "Frank", "Grace", "Heidi", "Ivan", "Judy", "Mallory", "Niaj"];
    const userAccs = {};
    users.forEach(u => {
        userAccs[u] = `ACC-${u.substring(0, 3).toUpperCase()}-${Math.floor(1000 + Math.random() * 9000)}`;
    });

    const txnId = (i) => `TXN-${10000 + i}`;

    // 1. Normal Background (50 transactions)
    for (let i = 0; i < 50; i++) {
        const sender = users[Math.floor(Math.random() * users.length)];
        const receiver = users.filter(u => u !== sender)[Math.floor(Math.random() * (users.length - 1))];
        const amount = (10 + Math.random() * 490).toFixed(2);
        const date = new Date(start_date.getTime() + i * 3600000);
        rows.push([
            txnId(rows.length), userAccs[receiver], receiver, "Transfer",
            date.toISOString().replace('T', ' ').substring(0, 19),
            amount, "10000", `M-${Math.floor(100 + Math.random() * 899)}`, "New York",
            sender, `U-${sender.substring(0, 3).toUpperCase()}`, userAccs[sender]
        ].join(','));
    }

    // 2. Structural Risk: Loop (Loop_1 -> Loop_2 -> Loop_3 -> Loop_4 -> Loop_1)
    const loopUsers = ["Loop1", "Loop2", "Loop3", "Loop4"];
    const loopAccs = {
        "Loop1": "L-ACC-100", "Loop2": "L-ACC-200", "Loop3": "L-ACC-300", "Loop4": "L-ACC-400"
    };
    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < loopUsers.length; j++) {
            const s = loopUsers[j];
            const r = loopUsers[(j + 1) % loopUsers.length];
            const date = new Date(start_date.getTime() + (50 * 3600000) + (i * 4 + j) * 600000);
            rows.push([
                txnId(rows.length), loopAccs[r], r, "Transfer",
                date.toISOString().replace('T', ' ').substring(0, 19),
                "500", "5000", "M-LOOP", "London",
                s, `U-${s.toUpperCase()}`, loopAccs[s]
            ].join(','));
        }
    }

    // 3. Behavioral/ML Risk: Burst
    const burstUser = "BurstKing";
    const burstAcc = "ACC-BURST-99";
    for (let i = 0; i < 15; i++) {
        const receiver = users[Math.floor(Math.random() * users.length)];
        const date = new Date(start_date.getTime() + (2 * 24 * 3600000) + i * 120000);
        rows.push([
            txnId(rows.length), userAccs[receiver], receiver, "Urgent Transfer",
            date.toISOString().replace('T', ' ').substring(0, 19),
            "8500", "50000", "M-URGENT", "Berlin",
            burstUser, "U-BURST", burstAcc
        ].join(','));
    }

    // 4. Proximity Risk: Seed
    const megaSeed = "MegaSeed";
    const seedAcc = "ACC-SEED-001";
    const seedDate = new Date(start_date.getTime() + 3 * 24 * 3600000);
    rows.push([
        txnId(rows.length), seedAcc, megaSeed, "Bulk Deposit",
        seedDate.toISOString().replace('T', ' ').substring(0, 19),
        "500000", "1000000", "M-OFFSHORE", "Cayman Islands",
        "ExternalSource", "EXT-001", "ACC-EXT-001"
    ].join(','));

    const brokers = ["BrokerA", "BrokerB", "BrokerC"];
    const brokerAccs = { "BrokerA": "ACC-BRK-1", "BrokerB": "ACC-BRK-2", "BrokerC": "ACC-BRK-3" };
    brokers.forEach((b, idx) => {
        const date = new Date(seedDate.getTime() + (idx + 1) * 3600000);
        rows.push([
            txnId(rows.length), brokerAccs[b], b, "Transfer",
            date.toISOString().replace('T', ' ').substring(0, 19),
            "50000", "60000", "M-SEED", "Tokyo",
            megaSeed, "U-SEED", seedAcc
        ].join(','));
    });

    brokers.forEach((b, idx) => {
        const target = users[Math.floor(Math.random() * users.length)];
        const date = new Date(seedDate.getTime() + (idx + 4) * 3600000);
        rows.push([
            txnId(rows.length), userAccs[target], target, "Payment",
            date.toISOString().replace('T', ' ').substring(0, 19),
            "2000", "15000", "M-BRK", "Tokyo",
            b, `U-${b.toUpperCase()}`, brokerAccs[b]
        ].join(','));
    });

    // 5. Mixed: Mule
    const mule = "MuleMike";
    const muleAcc = "ACC-MULE-404";
    for (let i = 0; i < 8; i++) {
        const date = new Date(start_date.getTime() + 4 * 24 * 3600000 + i * 3600000);
        // In
        const sender = users[Math.floor(Math.random() * users.length)];
        rows.push([
            txnId(rows.length), muleAcc, mule, "Transfer",
            date.toISOString().replace('T', ' ').substring(0, 19),
            "2500", "3000", "M-MULE-IN", "Madrid",
            sender, `U-${sender.substring(0, 3).toUpperCase()}`, userAccs[sender]
        ].join(','));
        // Out
        const target = users[Math.floor(Math.random() * users.length)];
        rows.push([
            txnId(rows.length), userAccs[target], target, "Transfer",
            new Date(date.getTime() + 1800000).toISOString().replace('T', ' ').substring(0, 19),
            "2450", "100", "M-MULE-OUT", "Madrid",
            mule, "U-MULE", muleAcc
        ].join(','));
    }

    fs.writeFileSync(filename, rows.join('\n'));
    console.log(`Generated ${rows.length - 1} transactions in ${filename}`);
}

generateDataset();

File: Miniproject-main/Miniproject-main/backend/data/generate_expanded_data.py
import random
import csv
import datetime

def generate_comprehensive_dataset(filename="comprehensive_transactions_100.csv"):
    headers = [
        "TransactionID", "AccountID", "AccountHolder", "TransactionType", 
        "TransactionDate", "Amount", "AccountBalance", "MerchantID", 
        "Location", "SenderName", "SenderID", "SenderAccountNo"
    ]
    rows = []
    
    start_date = datetime.datetime(2023, 11, 1, 9, 0, 0)
    
    # helper to generate IDs
    def txn_id(i): return f"TXN-{10000 + i}"
    def acc_id(name): return f"ACC-{name.upper()[:3]}-{random.randint(1000, 9999)}"
    
    users = ["Alice", "Bob", "Charlie", "David", "Eve", "Frank", "Grace", "Heidi", "Ivan", "Judy", "Mallory", "Niaj"]
    user_accs = {u: acc_id(u) for u in users}
    
    # 1. Normal Background (50 transactions) - Low Risk
    for i in range(50):
        sender = random.choice(users)
        receiver = random.choice([u for u in users if u != sender])
        amount = round(random.uniform(10, 500), 2)
        rows.append([
            txn_id(len(rows)), user_accs[receiver], receiver, "Transfer",
            (start_date + datetime.timedelta(hours=i)).strftime("%Y-%m-%d %H:%M:%S"),
            amount, 10000, f"M-{random.randint(100, 999)}", "New York",
            sender, f"U-{sender.upper()[:3]}", user_accs[sender]
        ])

    # 2. Structural Risk: Large Loop (Loop_1 -> Loop_2 -> Loop_3 -> Loop_4 -> Loop_1) (12 transactions)
    loop_users = ["Loop_1", "Loop_2", "Loop_3", "Loop_4"]
    loop_accs = {u: f"L-ACC-{u[-1]}00" for u in loop_users}
    for i in range(3): # repeat 3 times
        for j in range(len(loop_users)):
            s = loop_users[j]
            r = loop_users[(j + 1) % len(loop_users)]
            rows.append([
                txn_id(len(rows)), loop_accs[r], r, "Transfer",
                (start_date + datetime.timedelta(days=1, minutes=len(rows)*10)).strftime("%Y-%m-%d %H:%M:%S"),
                450, 5000, "M-LOOP", "London",
                s, f"U-{s.upper()}", loop_accs[s]
            ])

    # 3. Behavioral/ML Risk: High Velocity Burst (15 transactions)
    # Burst_User sends 15 transactions within 30 minutes
    burst_user = "Burst_King"
    burst_acc = "ACC-BURST-99"
    for i in range(15):
        receiver = random.choice(users)
        rows.append([
            txn_id(len(rows)), user_accs[receiver], receiver, "Urgent Transfer",
            (start_date + datetime.timedelta(days=2, minutes=i*2)).strftime("%Y-%m-%d %H:%M:%S"),
            8000, 50000, "M-URGENT", "Berlin",
            burst_user, "U-BURST", burst_acc
        ])

    # 4. Proximity Risk: High Value Seed (10 transactions downstream)
    # Mega_Seed (Very High Risk) -> Broker_1 -> Broker_2 -> Victim_1...
    mega_seed = "Mega_Seed"
    seed_acc = "ACC-SEED-001"
    # The infection source
    rows.append([
        txn_id(len(rows)), seed_acc, mega_seed, "Bulk Deposit",
        (start_date + datetime.timedelta(days=3)).strftime("%Y-%m-%d %H:%M:%S"),
        500000, 1000000, "M-OFFSHORE", "Cayman Islands",
        "External_Source", "EXT-001", "ACC-EXT-001"
    ])
    
    brokers = ["Broker_A", "Broker_B", "Broker_C"]
    broker_accs = {b: f"ACC-BRK-{b[-1]}" for b in brokers}
    
    # Seed to Brokers
    for b in brokers:
        rows.append([
            txn_id(len(rows)), broker_accs[b], b, "Transfer",
            (start_date + datetime.timedelta(days=3, hours=1)).strftime("%Y-%m-%d %H:%M:%S"),
            25000, 30000, "M-SEED", "Tokyo",
            mega_seed, "U-SEED", seed_acc
        ])
    
    # Brokers to legitimate-looking users
    for b in brokers:
        target = random.choice(users)
        rows.append([
            txn_id(len(rows)), user_accs[target], target, "Payment",
            (start_date + datetime.timedelta(days=3, hours=2)).strftime("%Y-%m-%d %H:%M:%S"),
            2000, 15000, "M-BRK", "Tokyo",
            b, f"U-{b.upper()}", broker_accs[b]
        ])

    # 5. Mixed Risk: "Mule" account (High in/out ratio) (12 transactions)
    mule = "Mule_Mike"
    mule_acc = "ACC-MULE-404"
    for i in range(6):
        # High incoming
        sender = random.choice(users)
        rows.append([
            txn_id(len(rows)), mule_acc, mule, "Transfer",
            (start_date + datetime.timedelta(days=4, hours=i)).strftime("%Y-%m-%d %H:%M:%S"),
            1500, 2000, "M-MULE-IN", "Madrid",
            sender, f"U-{sender.upper()[:3]}", user_accs[sender]
        ])
        # High outgoing
        target = random.choice(users)
        rows.append([
            txn_id(len(rows)), user_accs[target], target, "Transfer",
            (start_date + datetime.timedelta(days=4, hours=i, minutes=30)).strftime("%Y-%m-%d %H:%M:%S"),
            1450, 100, "M-MULE-OUT", "Madrid",
            mule, "U-MULE", mule_acc
        ])

    with open(filename, 'w', newline='') as f:
        writer = csv.writer(f)
        writer.writerow(headers)
        writer.writerows(rows)
    
    print(f"Dataset generated: {filename} with {len(rows)} transactions.")

if __name__ == "__main__":
    generate_comprehensive_dataset()
